File: 375.guess-number-higher-or-lower-ii.py
#
# @lc app=leetcode id=375 lang=python3
#
# [375] Guess Number Higher or Lower II
#

# Time:  O(n^2)
# Space: O(n^2)

# @lc code=start
class Solution:
    def getMoneyAmount(self, n: int) -> int:
        dp = [[0] * n for _ in range(n + 1)]
        for i in range(n - 1, -1, -1):
            for j in range(i + 1, n):
                dp[i][j] = min(k + 1 + max(dp[i][k - 1], dp[k + 1][j]) for k in range(i, j + 1))
        return dp[0][n - 1]


# @lc code=end

File: 375.guess-number-higher-or-lower-ii.js
/*
 * @lc app=leetcode id=375 lang=javascript
 *
 * [375] Guess Number Higher or Lower II
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var getMoneyAmount = function (n) {
  const dp = Array(n + 1)
    .fill()
    .map(() => Array(n + 1).fill(0));
  for (let i = 2; i < n + 1; ++i) {
    for (let j = i - 1; j > 0; --j) {
      let globalMin = Infinity;
      for (let k = j + 1; k < i; ++k) {
        let localMax = k + Math.max(dp[j][k - 1], dp[k + 1][i]);
        globalMin = Math.min(globalMin, localMax);
      }
      dp[j][i] = j === i - 1 ? j : globalMin;
    }
  }
  return dp[1][n];
};
// @lc code=end
